### PR TITLE
Declare Vault dependent projects

### PR DESCRIPTION
**Pull request by giorgiosironi**
For https://github.com/elifesciences/issues/issues/5002

If these projects do not have access to a readable Vault key for secrets, their highstate should fail hard.

### CODE REVIEW

**Comment by giorgiosironi** [comment]
Successfully throwing an exception from the `ext_pillar` implementation, but doesn't interrupt `salt-call pillar.get`:
```
2020-03-03 16:02:44,485 [salt.loaded.int.module.vault:148 ][ERROR   ][13742] Failed to read secret! HTTPError: 404 Client Error: Not Found for url: https://pr-14-fresh-manual--master-server.elifesciences.org:8200/v1/secret/data/projects/basebox/ci
2020-03-03 16:02:44,486 [salt.loaded.ext.pillar.elife_vault:54  ][WARNING ][13742] Error accessing Vault (basebox) in dependent project <class 'salt.exceptions.CommandExecutionError'>: 404 Client Error: Not Found for url: https://pr-14-fresh-manual--master-server.elifesciences.org:8200/v1/secret/data/projects/basebox/ci
2020-03-03 16:02:44,488 [salt.pillar      :966 ][ERROR   ][13742] Execption caught loading ext_pillar 'elife_vault':
  File "/usr/lib/python3/dist-packages/salt/pillar/__init__.py", line 956, in ext_pillar
    key)
  File "/usr/lib/python3/dist-packages/salt/pillar/__init__.py", line 875, in _external_pillar_data
    ext = self.ext_pillars[key](self.minion_id, pillar, **val)
  File "/opt/salt-extension-modules/pillar/elife_vault.py", line 55, in ext_pillar
    raise e
  File "/opt/salt-extension-modules/pillar/elife_vault.py", line 51, in ext_pillar
    vault_value = __salt__['vault.read_secret'](vault_key)
  File "/usr/lib/python3/dist-packages/salt/modules/vault.py", line 149, in read_secret
    raise salt.exceptions.CommandExecutionError(err)

2020-03-03 16:02:44,488 [salt.pillar      :1018][CRITICAL][13742] Pillar render error: Failed to load ext_pillar elife_vault: 404 Client Error: Not Found for url: https://pr-14-fresh-manual--master-server.elifesciences.org:8200/v1/secret/data/projects/basebox/ci
...
elife@ip-10-0-2-199:~$ sudo salt-call pillar.get elife:answer
local:
    42  # fallback to normal pillars
```

**Comment by giorgiosironi** [comment]
https://github.com/saltstack/salt/issues/23402#issuecomment-223473681 (infuriatingly) says to do this on the minion. I'll put in `os.exec("shutdown -h now")` if I have to.

**Comment by giorgiosironi** [comment]
Following the call stack, this is what the minion can see about the error:
```
elife@ip-10-0-2-199:~$ sudo salt-call pillar.get _errors
local:
    - Failed to load ext_pillar elife_vault: 404 Client Error: Not Found for url: https://pr-14-fresh-manual--master-server.elifesciences.org:8200/v1/secret/data/projects/basebox/ci
```
We could put in a state in `elife.init` that fails if this pillar is not empty. Still wouldn't stop _the rest_ of the states from running.

**Comment by giorgiosironi** [comment]
Found out that `pillar.get` does not fail, but highstate actually does in these conditions:
```
elife@ip-10-0-2-199:~$ sudo salt-call state.highstate
[INFO    ] Loading fresh modules for state activity
local:
    Data failed to compile:
----------
    Pillar failed to render with the following messages:
----------
    Failed to load ext_pillar elife_vault: 404 Client Error: Not Found for url: https://pr-14-fresh-manual--master-server.elifesciences.org:8200/v1/secret/data/projects/basebox/ci
```

**Comment by giorgiosironi** [comment]
If merged, this would only be enabled on [`basebox`](https://github.com/elifesciences/builder-configuration/pull/125).

That shouldn't fail, it only uses the `ci` environment:
```
elife@2018-04-09-2--master-server:~$ vault kv get secret/projects/basebox/ci
====== Metadata ======
Key              Value
---              -----
created_time     2019-02-13T16:49:10.500152114Z
deletion_time    n/a
destroyed        false
version          2

======== Data ========
Key              Value
---              -----
answer.number    43
```

**Comment by giorgiosironi** [comment]
`pr-base` is failing though.

**Comment by giorgiosironi** [comment]
Accidentally committed https://github.com/elifesciences/master-server-formula/commit/b6bd643be3b1a3bb6c4371d42fe601127a300ac9 to the `master` branch here (for real).

**Comment by giorgiosironi** [comment]
Testing https://github.com/elifesciences/master-server-formula/pull/15 first.

File: salt/master-server/config/opt-salt-extension-modules-pillar-elife_vault.py [hunk 1]
```diff
@@ -37,18 +38,24 @@ def ext_pillar(minion_id, pillar, path=None, env_key=None):
...
     log.info("Reading vault_key: %s", vault_key)
     try:
         vault_value = __salt__['vault.read_secret'](vault_key)
     except Exception as e:
-        log.warning("Error accessing Vault (%s): %s", type(e), str(e))
-        return {}
+        if project in dependent_projects:
+            log.warning("Error accessing Vault (%s) in dependent project %s: %s", project, type(e), str(e))
+            raise e
+        else:
```
Comment (thewilkybarkid) [inline salt/master-server/config/opt-salt-extension-modules-pillar-elife_vault.py]: Condition isn’t needed as the `if` raises.

File: salt/master-server/config/opt-salt-extension-modules-pillar-elife_vault.py [hunk 2]
```diff
@@ -37,18 +38,24 @@ def ext_pillar(minion_id, pillar, path=None, env_key=None):
...
         log.warn("Stack %s does not have a grain: %s", __grains__.get('id'), e)
         return {}
     log.info("Reading vault_key: %s", vault_key)
     try:
         vault_value = __salt__['vault.read_secret'](vault_key)
     except Exception as e:
-        log.warning("Error accessing Vault (%s): %s", type(e), str(e))
-        return {}
+        if project in dependent_projects:
+            log.warning("Error accessing Vault (%s) in dependent project %s: %s", project, type(e), str(e))
```
Comment (thewilkybarkid) [inline salt/master-server/config/opt-salt-extension-modules-pillar-elife_vault.py]: Assume there’s a higher level that should be used?

File: salt/master-server/config/opt-salt-extension-modules-pillar-elife_vault.py [hunk 3]
```diff
@@ -37,18 +38,24 @@ def ext_pillar(minion_id, pillar, path=None, env_key=None):
...
+            raise e
+        else:
+            log.warning("Error accessing Vault (%s): %s", type(e), str(e))
```
Comment (thewilkybarkid) [inline salt/master-server/config/opt-salt-extension-modules-pillar-elife_vault.py]: Worth rewording to say it’s carrying on anyway?

File: salt/master-server/config/opt-salt-extension-modules-pillar-elife_vault.py (hunk 2 above)
Comment (giorgiosironi) [reply salt/master-server/config/opt-salt-extension-modules-pillar-elife_vault.py]: Logged at an higher level when the whole pillar rendering fails, but will switch it here too.

File: salt/master-server/config/opt-salt-extension-modules-pillar-elife_vault.py (hunk 1 above)
Comment (giorgiosironi) [reply salt/master-server/config/opt-salt-extension-modules-pillar-elife_vault.py]: Will make the `else` the base case.

**Comment by giorgiosironi** [comment]
> Is it not possible to know which projects need Vault and which ones don’t

Could maybe switch the pillars to specify whether Vault is used rather than whether to error out, but it's more difficult to rollout as now I have put only `basebox` in the pillars and plan to add one project at a time.

**Comment by giorgiosironi** [comment]
Let's try this, rolling it out slowly.

**Pull request merged**